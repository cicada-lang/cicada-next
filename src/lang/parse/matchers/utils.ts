import * as Exps from "../../exp"

export function piBindingtoFnBindingFrom(binding: Exps.PiBinding): Exps.FnBinding {
  switch (binding.kind) {
    case "PiBindingNameless": {
      return Exps.FnBindingName("_")
    }

    case "PiBindingNamed": {
      return Exps.FnBindingName(binding.name)
    }

    case "PiBindingImplicit": {
      return Exps.FnBindingImplicit(binding.name)
    }
  }
}
