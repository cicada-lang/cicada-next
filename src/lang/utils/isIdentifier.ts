const identifierRegExp = new RegExp("^([-_\\p{Letter}][-_\\p{Letter}0-9]*)$", "u")

export function isIdentifier(text: string): boolean {
  return identifierRegExp.test(text)
}
