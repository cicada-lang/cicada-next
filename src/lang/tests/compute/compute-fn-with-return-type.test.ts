import { expect, test } from "vitest"
import { expectCodeToFail, runCode } from "../utils"

test("compute Fn with return type", async () => {
  const output = await runCode(`

let id = function (T: Type, x: T): T { return x }
compute id
compute id(Type)

`)

  expect(output).toMatchInlineSnapshot(`
    "(T2, x11) => x11: (T2: Type, x11: T2) -> T2
    (x2) => x2: (x2: Type) -> Type"
  `)
})

test("compute Fn with return type -- conflict", async () => {
  const output = await expectCodeToFail(`

let id = function (x: String): Type { return x }

  `)
})

test("compute Fn return type -- stmts", async () => {
  const output = await runCode(`

function id(T: Type, x: T): T { 
  return x 
}

compute id

`)

  expect(output).toMatchInlineSnapshot(
    '"(T2, x11) => x11: (T2: Type, x11: T2) -> T2"',
  )
})
