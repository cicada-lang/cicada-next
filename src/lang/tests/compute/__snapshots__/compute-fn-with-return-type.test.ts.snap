// Vitest Snapshot v1, https://bun:test.dev/guide/snapshot.html

exports[`compute Fn return type -- stmts 1`] = `"(T, x) => x: (T: Type, T) -> T"`;

exports[`compute Fn with return type 1`] = `
"(T, x) => x: (T: Type, T) -> T
(x) => x: (Type) -> Type"
`;
